src/Spec.ts

```diff
@@ -539,6 +548,240 @@ export default class Spec {
+            sublist: tree.contents,
+          } as OrderedListItemNode;
+          for (const step of path) {
+            stepPointer = stepPointer.sublist!.contents[step];
+          }
+          this.warn({
+            type: 'contents',
+            node: alg,
+            ruleId: 'invocation-return-type',
+            message,
+            nodeRelativeLine: stepPointer.location.start.line,
+            nodeRelativeColumn: stepPointer.location.start.column,
+          });
+        }
+      };
+
+      const consumedAsCompletion = isConsumedAsCompletion(xref);
+      // checks elsewhere ensure that well-formed documents never have a union of completion and non-completion, so checking the first child suffices
+      // TODO: this is for 'a break completion or a throw completion', which is kind of a silly union; maybe address that in some other way?
+      const isCompletion =
+        returnType.kind === 'completion' ||
+        (returnType.kind === 'union' && returnType.types[0].kind === 'completion');
+      if (['Completion', 'ThrowCompletion', 'NormalCompletion'].includes(xref.aoid)) {
+        if (consumedAsCompletion) {
+          // TODO mf to improve message
+          warn(
+            `${xref.aoid} is explicitly a Completion Record and should not be unwrapped or marked`
+          );
+        }
+      } else if (isCompletion && !consumedAsCompletion) {
+        warn(`${xref.aoid} returns a Completion Record, but is not consumed as one`);
+      } else if (!isCompletion && consumedAsCompletion) {
+        warn(`${xref.aoid} does not return a Completion Record, but is consumed as one`);
+      }
+      if (returnType.kind === 'unused' && !isCalledAsPerform(xref, false)) {
+        warn(
+          `${xref.aoid} does not return a value and should only be invoked as \`Perform ${xref.aoid}(...).\``
+        );
+      }
+
+      if (onlyPerformed.has(xref.aoid) && onlyPerformed.get(xref.aoid) !== 'top') {
+        const old = onlyPerformed.get(xref.aoid);
+        const performed = isCalledAsPerform(xref, true);
+        if (!performed) {
+          onlyPerformed.set(xref.aoid, 'top');
+        } else if (old === null) {
+          onlyPerformed.set(xref.aoid, 'only performed');
+        }
+      }
+      if (
+        alwaysAssertedToBeNormal.has(xref.aoid) &&
+        alwaysAssertedToBeNormal.get(xref.aoid) !== 'top'
+      ) {
+        const old = alwaysAssertedToBeNormal.get(xref.aoid);
+        const asserted = isAssertedToBeNormal(xref);
+        if (!asserted) {
+          alwaysAssertedToBeNormal.set(xref.aoid, 'top');
+        } else if (old === null) {
+          alwaysAssertedToBeNormal.set(xref.aoid, 'always asserted normal');
+        }
+      }
+    }
+
+    for (const [aoid, state] of onlyPerformed) {
+      if (state !== 'only performed') {
+        continue;
+      }
+      const message = `${aoid} is only ever invoked with Perform, so it should return ~unused~ or a Completion Record normally containing ~unused~`;
```

Comment:
```suggestion
      const message = `${aoid} is only ever invoked with Perform, so it should return ~unused~ or a normal completion containing ~unused~`;
```

Side note: it should be okay to always use perform with an AO that always returns an abrupt completion. We could tweak `isUnused` above to allow for that, but we probably don't need this message to mention that case.

Reply:
It's also ok to use it with an AO that returns either a normal completion containing unused _or an abrupt completion_. Changed to `... a Completion Record which, if normal, contains ~unused~`.